src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/ExtensionMethodFilteringService.cs

```diff
@@ -0,0 +1,191 @@
+                            continue;
+                        }
+
+                        var info = await document.GetSyntaxTreeIndexAsync(loadOnly, cancellationToken).ConfigureAwait(false);
+
+                        // Don't provide anyting if we don't have all the required SyntaxTreeIndex created.
+                        if (info == null)
+                        {
+                            return null;
+                        }
+
+                        if (info.ContainsExtensionMethod)
+                        {
+                            syntaxTreeIndexBuilder.Add(info);
+                        }
+                    }
+                }
+
+                foreach (var peReference in peReferences)
+                {
+                    var info = await SymbolTreeInfo.GetInfoForMetadataReferenceAsync(
+                        solution, peReference, loadOnly, cancellationToken).ConfigureAwait(false);
+
+                    // Don't provide anyting if we don't have all the required SymbolTreeInfo created.
+                    if (info == null)
+                    {
+                        return null;
+                    }
+
+                    if (info.ContainsExtensionMethod)
+                    {
+                        symbolTreeInfoBuilder.Add(info);
+                    }
+                }
+
+                var results = new MultiDictionary<string, string>();
+
+                // Find matching extension methods from source.
+                foreach (var info in syntaxTreeIndexBuilder)
+                {
+                    // Add simple extension methods with matching target type name
+                    foreach (var targetTypeName in targetTypeNames)
+                    {
+                        if (!info.SimpleExtensionMethodInfo.TryGetValue(targetTypeName, out var methodInfoIndices))
+                        {
+                            continue;
+                        }
+
+                        foreach (var index in methodInfoIndices)
+                        {
+                            if (info.TryGetDeclaredSymbolInfo(index, out var methodInfo))
+                            {
+                                results.Add(methodInfo.FullyQualifiedContainerName, methodInfo.Name);
+                            }
+                        }
+                    }
+
+                    // Add all complex extension methods, we will need to completely rely on symbols to match them.
+                    foreach (var index in info.ComplexExtensionMethodInfo)
+                    {
+                        if (info.TryGetDeclaredSymbolInfo(index, out var methodInfo))
+                        {
+                            results.Add(methodInfo.FullyQualifiedContainerName, methodInfo.Name);
+                        }
+                    }
+                }
+
+                // Find matching extension methods from metadata
+                foreach (var info in symbolTreeInfoBuilder)
```

Comment:
if you remove the quick-bail checks, then you don't need these temp constructs to loop over.

Reply:
I'm inclined to still do the quick bail, i.e. not return partial results.